### PR TITLE
[ORTModule] ATen Support for aten::upsample_nearest

### PR DESCRIPTION
**Pull request by iK1D**
ATen support for aten::upsample_nearest, which is required for Huggingface's diffusers model training using ORTModule.

### CODE REVIEW

File: onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc [hunk 1]
```diff
@@ -78,22 +60,26 @@ struct ATenOperator {
...
                               dlpack->dl_tensor.dtype.bits == 64);
-        i_value = is_list ? ToListIValue<int64_t>(dlpack, is_optional) : ToIValue<int64_t>(dlpack, is_optional);
+        i_value = is_list ? ToListIValue<int64_t, int64_t>(dlpack, is_optional)
+                          : ToIValue<int64_t, int64_t>(dlpack, is_optional);
       } break;
       case c10::TypeKind::FloatType: {
         TORCH_INTERNAL_ASSERT(dlpack->dl_tensor.dtype.code == DLDataTypeCode::kDLFloat &&
                               dlpack->dl_tensor.dtype.bits == 32);
-        i_value = is_list ? ToListIValue<float>(dlpack, is_optional) : ToIValue<float>(dlpack, is_optional);
+        // PyTorch's IValue doesn't support float, so we convert it to double.
```
Comment (baijumeswani) [inline onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:69]: Does pytorch perform computation with doubles for non-tensor `IValue`s?

File: onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc [hunk 2]
```diff
@@ -6,42 +6,24 @@
...
 
-template <typename T>
+template <typename TSrc, typename TDst>
 c10::IValue ToIValue(const DLManagedTensor* dlpack, bool is_optional) {
   TORCH_INTERNAL_ASSERT((dlpack->dl_tensor.ndim == 0 && dlpack->dl_tensor.shape == nullptr) ||
                         (dlpack->dl_tensor.ndim == 1 && dlpack->dl_tensor.shape[0] == 1));
-  T value = *reinterpret_cast<const T*>(dlpack->dl_tensor.data);
-  return is_optional ? c10::IValue(c10::optional<T>(value)) : c10::IValue(value);
+  TDst value = static_cast<TDst>(*reinterpret_cast<const TSrc*>(dlpack->dl_tensor.data));
+  return is_optional ? c10::IValue(c10::optional<TDst>(value)) : c10::IValue(value);
```
Comment (baijumeswani) [inline onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:14]: Was `c10::IValue(value)` performing an implicit cast until now? Why is the change required now (does the implicit cast no longer work)?

File: onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc (hunk 1 above)
Comment (iK1D) [reply onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:69]: The computation here for upsample_nearest still take float list, but we need to pass arguments by IValue to them. PyTorch will do the float to double conversion when calling the function.

File: onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc (hunk 2 above)
Comment (iK1D) [reply onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:14]: IValue(double) can take float as input, but IValue(c10::List) cannot take c10::List<float>. That's why I need this change.

File: onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc (hunk 1 above)
Comment (baijumeswani) [reply onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:69]: > PyTorch will do the float to double conversion when calling the function.

you mean double to float right?

**Review by baijumeswani — APPROVED** [review]

File: onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc (hunk 1 above)
Comment (iK1D) [reply onnxruntime/python/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:69]: Right...

**Pull request merged**